clients/drcachesim/scheduler/scheduler.cpp

```diff
@@ -1547,21 +1548,38 @@ scheduler_tmpl_t<RecordType, ReaderType>::start_speculation(output_ordinal_t out
                                                             addr_t start_address,
                                                             bool queue_current_record)
 {
-    if (!outputs_[output].speculating && queue_current_record) {
-        inputs_[outputs_[output].cur_input].queue.push_back(outputs_[output].last_record);
+    auto &outinfo = outputs_[output];
+    if (outinfo.speculation_split.empty()) {
+        if (queue_current_record)
+            inputs_[outinfo.cur_input].queue.push_back(outinfo.last_record);
+        outinfo.speculation_split.push(0);
+    } else {
+        if (queue_current_record) {
+            // XXX: We'll re-call the speculator so we're assuming a repeatable response.
```

Comment:
You mean we'll get the same response from the speculator when we re-start at prev_speculate_pc? I don't know how the speculator is implemented so not sure if it depends on any other state. But does it matter if the speculator returns a different pc after resumption at prev_speculate_pc?

Reply:
If queuing of the already-seen instruction was requested at start_speculation, the speculator has to return that same instruction in all details, not just the next pc.  Maybe it should be saved -- the speculator could always remember every start-request instruction, or maybe the scheduler should add a saved record to this stack.

Reply:
For now I will put a TODO both here and in the speculator.  This won't be an issue until we do more than return a nop.

Reply:
Done.